process_openbsd.go

```diff
@@ -0,0 +1,359 @@
+	Ki_ustart_usec  uint32
+	Ki_uutime_sec   uint32
+	Ki_uutime_usec  uint32
+	Ki_ustime_sec   uint32
+	Ki_ustime_usec  uint32
+	Ki_pad_cgo_0    [4]byte
+	Ki_uru_maxrss   uint64
+	Ki_uru_ixrss    uint64
+	Ki_uru_idrss    uint64
+	Ki_uru_isrss    uint64
+	Ki_uru_minflt   uint64
+	Ki_uru_majflt   uint64
+	Ki_uru_nswap    uint64
+	Ki_uru_inblock  uint64
+	Ki_uru_oublock  uint64
+	Ki_uru_msgsnd   uint64
+	Ki_uru_msgrcv   uint64
+	Ki_uru_nsignals uint64
+	Ki_uru_nvcsw    uint64
+	Ki_uru_nivcsw   uint64
+	Ki_uctime_sec   uint32
+	Ki_uctime_usec  uint32
+	Ki_psflags      int32
+	Ki_spare        int32
+	Ki_svuid        uint32
+	Ki_svgid        uint32
+	Ki_emul         [8]int8
+	Ki_rlim_rss_cur uint64
+	Ki_cpuid        uint64
+	Ki_vm_map_size  uint64
+	Ki_tid          int32
+	Ki_rtableid     uint32
+}
+
+var proc_k_size = unsafe.Sizeof(Kinfo_proc{})
+
+// UnixProcess is an implementation of Process that contains Unix-specific
+// fields and information.
+type UnixProcess struct {
+	pid   int
+	ppid  int
+	state rune
+	pgrp  int
+	sid   int
+
+	binary string
+}
+
+// Pid returns process id
+func (p *UnixProcess) Pid() int {
+	return p.pid
+}
+
+// PPid returns parent process id
+func (p *UnixProcess) PPid() int {
+	return p.ppid
+}
+
+// Executable returns process executable name
+func (p *UnixProcess) Executable() string {
+	return p.binary
+}
+
+// Path returns path to process executable
+func (p *UnixProcess) Path() (string, error) {
+	// Adapted from aja's fix for osext:
+	// https://github.com/kardianos/osext/commit/b4814f465fb1f92d46e37f7ef84d732ece7c3e3a
+
+	var args []string
+	var execPath string
+	mib := []int32{CTL_KERN, KERN_PROC_ARGS, int32(p.pid), KERN_PROC_ARGV}
+	n := uintptr(0)
+
+	_, _, errNum := syscall.Syscall6(syscall.SYS___SYSCTL, uintptr(unsafe.Pointer(&mib[0])), 4, 0, uintptr(unsafe.Pointer(&n)), 0, 0)
+	if errNum != 0 {
+		return "", errNum
+	}
+	if n == 0 { // This shouldn't happen.
+		return "", nil
+	}
+
+	buf := make([]byte, n)
+	_, _, errNum = syscall.Syscall6(syscall.SYS___SYSCTL, uintptr(unsafe.Pointer(&mib[0])), 4, uintptr(unsafe.Pointer(&buf[0])), uintptr(unsafe.Pointer(&n)), 0, 0)
+	if errNum != 0 {
+		return "", errNum
+	}
+	if n == 0 { // This shouldn't happen.
+		return "", nil
+	}
+
+	argv := uintptr(unsafe.Pointer(&buf[0]))
+Loop:
+	for {
+		argp := *(**[1 << 20]byte)(unsafe.Pointer(argv))
+		if argp == nil {
+			break
+		}
+		for i := 0; uintptr(i) < n; i++ {
+			// we don't want the full arguments list
+			if string(argp[i]) == " " {
+				break Loop
+			}
+			if argp[i] != 0 {
+				continue
+			}
+			args = append(args, string(argp[:i]))
+			n -= uintptr(i)
+			break
+		}
+		if n < unsafe.Sizeof(argv) {
+			break
+		}
+		argv += unsafe.Sizeof(argv)
+		n -= unsafe.Sizeof(argv)
+	}
+	execPath = args[0]
+	// There is no canonical way to get an executable path on
+	// OpenBSD, so check PATH in case we are called directly
+	if execPath[0] != '/' && execPath[0] != '.' {
+		execIsInPath, err := exec.LookPath(execPath)
```

Comment:
Yeah, PATH has an impact on this, if you have a "godoc" binary in ~/bin and in /bin and your PATH=/bin:~/bin, it will return the one in /bin.

I can just put back the original return if you think it's better.